File: core/src/main/java/moe/caa/multilogin/core/ohc/LoggingInterceptor.java
package moe.caa.multilogin.core.ohc;

import moe.caa.multilogin.api.internal.logger.LoggerProvider;
import okhttp3.*;
import okio.Buffer;
import okio.BufferedSource;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

/**
 * Http 日志打印拦截器
 */
public class LoggingInterceptor implements Interceptor {

    @NotNull
    @Override
    public Response intercept(@NotNull Chain chain) throws IOException {
        Request request = chain.request();
        LoggerProvider.getLogger().debug(String.format("--> %s %s", request.method(), request.url()));

        RequestBody requestBody = request.body();
        if (requestBody != null) {
            Buffer bf = new Buffer();
            requestBody.writeTo(bf);
            long size = bf.size();
            if (size > 0) LoggerProvider.getLogger().debug(String.format("--> (%d bytes)", size));
        }

        long startNs = System.nanoTime();
        Response response;
        try {
            response = chain.proceed(request);
        } catch (Exception e) {
            LoggerProvider.getLogger().debug("<-- HTTP FAILED", e);
            throw e;
        }
        long tookMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNs);

        LoggerProvider.getLogger().debug(String.format("<-- %s %s (%dms)", response.code(), response.request().url(), tookMs));
        ResponseBody body = response.body();
        if (body != null) {

            BufferedSource source = body.source();
            source.request(Long.MAX_VALUE);
            Buffer buffer = source.getBuffer();
            long size = buffer.size();
            if (size > 0) LoggerProvider.getLogger().debug(String.format("<-- (%d bytes)", size));
        }

        return response;
    }
}

File: core/src/main/kotlin/moe/caa/multilogin/core/intercept/LoggingInterceptor.kt
package moe.caa.multilogin.core.intercept

import moe.caa.multilogin.api.logger.logDebug
import okhttp3.Interceptor
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.Response
import okio.Buffer
import java.util.concurrent.TimeUnit

/**
 * Http 日志打印拦截器
 */
object LoggingInterceptor : Interceptor {
    override fun intercept(chain: Interceptor.Chain): Response {
        val request: Request = chain.request()
        val requestBody: RequestBody? = request.body

        logDebug("--> ${request.method} ${request.url}")
        if (requestBody != null) {
            val bf = Buffer()
            requestBody.writeTo(bf)
            val contentLength = bf.size
            if (contentLength > 0) {
                logDebug("--> ($contentLength bytes)")
                requestBody.toString()
            }
        }

        val startNs = System.nanoTime()
        val response: Response
        try {
            response = chain.proceed(request)
        } catch (e: Exception) {
            logDebug("<-- HTTP FAILED", e)
            throw e
        }
        val tookMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNs)

        logDebug("<-- ${response.code} ${response.request.url} (${tookMs}ms)")
        val body = response.body
        if (body != null) {
            body.source().request(Long.MAX_VALUE)
            val contentLength = body.source().buffer.size
            if (contentLength > 0) logDebug("<-- ($contentLength bytes)")
        }

        return response
    }
}